feat(web): add support for language packs

jsjoeio opened the pull request:
## Description

This PR adds support for using language packs with VS Code Web via the CLI flag `--locale`.

## Known Limitations

There are a couple limitations with this implementation. Hoping that the Code Team can provide guidance on whether these limitations need to be addressed before merging and if so, how they recommend addressing them. 

### Implementation only works on a full build 

I am not sure why exactly, but this patch only works when you do a full build of VS Code and run that (meaning you can't use this in dev mode).

### Requires `commit` in `product.json`

I had to manually add commit to `product.json` in the build output because of [this logic](https://github.com/microsoft/vscode/blob/main/src/vs/server/node/remoteLanguagePacks.ts#L22).

### Only works with `--locale` CLI flag

Switching the display language only works using the `--locale` CLI flag. I believe this is because the `argv.json` is only modified in the client and saved in the IndexedDB. 

![image](https://user-images.githubusercontent.com/3806031/160485113-9ef519af-d38b-4fc0-b017-a2d48cece71b.png)

If we patched [this](https://github.com/microsoft/vscode/blob/e16d9ef81c4740b6f0a6afec86b69583067de0a8/src/vs/workbench/services/environment/browser/environmentService.ts#L61) to use the file system instead of `userRoamingDataHome`, it could potentially work normally (without passing a CLI flag). 

### Still shows "Install in Browser" which leads to error

Even if it's installed on the remote/server, you still see the "Install in Browser" button. When you click on it, you get an error toast notification. Error comes from [this block](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/extensionManagement/common/webExtensionManagementService.ts#L64-L66). Do language packs not have manifests?

![image](https://user-images.githubusercontent.com/3806031/160883552-db36a816-55a3-434d-a2a9-16d4ea89fce7.png)

![image](https://user-images.githubusercontent.com/3806031/160883480-b9d535b1-d47b-47ad-8192-0d27a1cd1d32.png)


## Demo

https://user-images.githubusercontent.com/3806031/160885289-7b5ceb6e-8c1d-4e11-8fdf-b3933f4bae62.mov

## How to Test
1. `git clone https://github.com/microsoft/vscode.git` 
2. `gh pr checkout 146012`
3. `yarn`
4. `yarn gulp "vscode-reh-web-darwin-x64-min"` (replace `darwin-x64` with your platform)
5. `cd .. &&  ./vscode-reh-web-darwin-x64/bin/code-server-oss` 
6. Open VS Code in browser
7. Download an extension pack (i.e. [Spanish](https://marketplace.visualstudio.com/items?itemName=MS-CEINTL.vscode-language-pack-es))
8. Install from VSIX file
9. Kill VS Code and run again:  `./vscode-reh-web-darwin-x64/bin/code-server-oss --locale es` 
10. See display in Spanish

Fixes #82595
Fixes https://github.com/coder/code-server/issues/4703

## Contributors

- @code-asher - wrote original patch
- @jsjoeio - extracted from code-server

jsjoeio commented on src/vs/server/node/remoteLanguagePacks.ts:
```diff
@@ -44,3 +50,39 @@ export namespace InternalNLSConfiguration {
 	}
 }
+
+export const getLocaleFromConfig = async (argvResource: string): Promise<string> => {
```
May be able to use this function: https://github.com/microsoft/vscode/blob/f3a9245af988874dd1bae7281d904ff6133304c6/src/main.js#L585

jsjoeio replied:
Actually...not 1:1 and can't import from JS file I don't think (well I don't see any other files using this since it's not exported) so going to leave as is.

jsjoeio replied:
Similar function/logic as well: https://github.com/microsoft/vscode/blob/f3a9245af988874dd1bae7281d904ff6133304c6/src/main.js#L233-L256

jsjoeio commented on src/vs/server/node/remoteLanguagePacks.ts:
```diff
@@ -44,3 +50,39 @@ export namespace InternalNLSConfiguration {
+	}
+};
+
+const stripComments = (content: string): string => {
```
Looks like this is now exported and available here: https://github.com/microsoft/vscode/blob/f3a9245af988874dd1bae7281d904ff6133304c6/src/vs/base/common/json.ts#L1316

jsjoeio commented:
I'm running into issues building locally. Any ideas?

```typescript
[15:28:58] Error: /Users/jp/dev/oss/vscode/src/vs/workbench/contrib/debug/node/terminals.ts(9,10): '"windows-process-tree"' has no exported member named 'IProcessTreeNode'. Did you mean 'ProcessTreeNode'?
[15:28:58] Error: /Users/jp/dev/oss/vscode/src/vs/platform/terminal/node/windowsShellHelper.ts(135,89): '"/Users/jp/dev/oss/vscode/node_modules/@types/windows-process-tree/index"' has no exported member named 'IProcessTreeNode'. Did you mean 'ProcessTreeNode'?
[15:28:58] Finished compilation with 2 errors after 165738 ms
[15:28:58] Error: /Users/jp/dev/oss/vscode/src/vs/workbench/contrib/debug/node/terminals.ts(9,10): '"windows-process-tree"' has no exported member named 'IProcessTreeNode'. Did you mean 'ProcessTreeNode'?
[15:28:58] Error: /Users/jp/dev/oss/vscode/src/vs/platform/terminal/node/windowsShellHelper.ts(135,89): '"/Users/jp/dev/oss/vscode/node_modules/@types/windows-process-tree/index"' has no exported member named 'IProcessTreeNode'. Did you mean 'ProcessTreeNode'?
[15:28:58] Finished compilation with 2 errors after 165740 ms
[15:28:58] 'vscode-reh-web-darwin-x64-min' errored after 2.85 min
[15:28:58] Error: Found 2 errors
    at Stream.<anonymous> (/Users/jp/dev/oss/vscode/build/lib/reporter.js:91:29)
    at _end (/Users/jp/dev/oss/vscode/node_modules/through/index.js:65:9)
    at Stream.stream.end (/Users/jp/dev/oss/vscode/node_modules/through/index.js:74:5)
    at Stream.onend (internal/streams/legacy.js:47:10)
    at Stream.emit (events.js:412:35)
    at Stream.emit (domain.js:532:15)
    at drain (/Users/jp/dev/oss/vscode/node_modules/through/index.js:34:23)
    at Stream.stream.queue.stream.push (/Users/jp/dev/oss/vscode/node_modules/through/index.js:45:5)
    at Stream.end (/Users/jp/dev/oss/vscode/node_modules/through/index.js:15:35)
    at _end (/Users/jp/dev/oss/vscode/node_modules/through/index.js:65:9)
error Command failed with exit code 1.
```

TylerLeonhardt commented:
Looks like the build is in a bad state. Unrelated to your change. Should be fixed tomorrow.

TylerLeonhardt commented:
@jsjoeio just an FYI I updated your brand from main to help with the CI and local failures.

TylerLeonhardt commented:
> Implementation only works on a full build
> Requires commit in product.json

Yes I believe all of this is expected and is the same as Desktop. The commit is used in the path of one of the localization files and also when running out of sources we don’t bundle files (for easier debugging) but bundling is important for the localization feature. @dbaeumer please let me know if I have this wrong.

TylerLeonhardt commented:
> Only works with --locale CLI flag

yeah I’d like to get to the bottom of this before we merge this PR in. I should have some cycles to look into it this week.

jsjoeio commented:
> I should have some cycles to look into it this week.

Awesome! Thank you so much for the quick replies and offering to help us get to the bottom of this! Really appreciate it. Let me know how else I can be helpful.

jsjoeio commented:
Okay a couple updates:
- added demo video
- updated testing instructions

I'm not sure how to fix this Hygine error:

```shell
$ node build/lib/layersChecker.js
[build/lib/layersChecker.ts]: Reference to 'document' from 'lib.dom.d.ts' violates layer '**/vs/base/common/platform.ts' (/home/runner/work/vscode/vscode/src/vs/base/common/platform.ts ([8](https://github.com/microsoft/vscode/runs/5756751534?check_suite_focus=true#step:11:8)7,20)
[build/lib/layersChecker.ts]: Reference to 'document' from 'lib.dom.d.ts' violates layer '**/vs/base/common/platform.ts' (/home/runner/work/vscode/vscode/src/vs/base/common/platform.ts (87,48)
[build/lib/layersChecker.ts]: Reference to 'getElementById' from 'lib.dom.d.ts' violates layer '**/vs/base/common/platform.ts' (/home/runner/work/vscode/vscode/src/vs/base/common/platform.ts (87,57)
[build/lib/layersChecker.ts]: Reference to 'getAttribute' from 'lib.dom.d.ts' violates layer '**/vs/base/common/platform.ts' (/home/runner/work/vscode/vscode/src/vs/base/common/platform.ts (88,32)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I see `navigator` used in the same block where I'm using `document`.

Maybe there's a way to refactor this code to do it the "proper way":
```typescript
	const el = typeof document !== 'undefined' && document.getElementById('vscode-remote-nls-configuration');
	const rawNlsConfig = el && el.getAttribute('data-settings');
```

jsjoeio commented:
Friendly bump @TylerLeonhardt - do you think you'll have a chance to look at this this month?

jsjoeio commented:
@TylerLeonhardt no worries at all - hope everything is okay personally! Thanks so much for the update. 

> I can speak more about that once I actually have something to show off 😅 but essentially the idea is to have navigator.language dictate the locale to use.

That sounds amazing! I'll keep an eye out then for an update from you 😄

jsjoeio commented:
@TylerLeonhardt now that 1.68 has [core localization support](https://code.visualstudio.com/updates/v1_68#_core-localization-support), is this still needed?

TylerLeonhardt commented:
I think we can close this now since we went in a different direction. This wasn't thrown away though, we all learned from the work you did for sure so thank you for taking the time to contribute. It means a lot to the community!
